### ISSUE TITLE
Wrong distance values?

### ISSUE THREAD

**Issue by Artif-Inteller** [js1010/cuhnsw#25]
I extracted `500 * 128` sized SIFT features of an image.
I built a `cuhnsw` index of that vector and queried with the same vector.
Desired distance values(`topk = 1`, `l2` distance) are all `0`, but I got very big values(greater than 10^6).
I tried to change some hyperparameters, but got the same result.

Hope help me with this issue.

Thanks

**Comment by js1010** [comment]
I think you misused `nrz` option. if `nrz = True`, library normalizes the training data (this option is for cosine similarity space by setting `dist_type = "dot"` and `nrz = True`). Unfortunately nrz is true by default.. see below permlinks

https://github.com/js1010/cuhnsw/blob/fbd41acd3235aab46533e12ad6b424775123a8d6/cuhnsw/proto/config.proto#L24
https://github.com/js1010/cuhnsw/blob/fbd41acd3235aab46533e12ad6b424775123a8d6/cuhnsw/pyhnsw.py#L42-L43

Also, I added function to example code to check validity of distance function

https://github.com/js1010/cuhnsw/pull/26

I got following results by shifting data from glove-50-angular to sift-128-euclidean in `example1.py`

```
# python example1.py run_gpu_inference2 --index-file=cuhnsw.index
====================================================================================================
[INFO    ] 2021-04-19 20:44:07 [example1.py] [run_gpu_inference2:180]gpu inference on res/sift-128-euclidean.hdf5 with index res/cuhnsw.index
[INFO    ] 2021-04-19 20:44:07 [pyhnsw.py] [__init__:30]opt: {
  "c_log_level": 2,
  "ef_construction": 150,
  "level_mult": 0.40242960438184466,
  "hyper_threads": 100.0,
  "block_dim": 32,
  "dist_type": "l2",
  "nrz": false,
  "reverse_cand": false,
  "heuristic_coef": 0.0,
  "seed": 777,
  "py_log_level": 2,
  "max_m": 12,
  "max_m0": 24,
  "save_remains": false,
  "visited_table_size": 0,
  "visited_list_size": 8192
}
[info    ] 2021-04-19 20:44:07 [cuhnsw_base.cu:59] cuda device info, major: 7, minor: 5, multi processors: 40, cores: 2560
[INFO    ] 2021-04-19 20:44:07 [example1.py] [run_gpu_inference2:182]load model from res/cuhnsw.index by cuhnsw
====================================================================================================
query 1
----------------------------------------------------------------------------------------------------
rank 1. neighbor: 932085, dist by lib: 232.8712158203125, actual dist: 232.8712158203125
rank 2. neighbor: 934876, dist by lib: 234.7147216796875, actual dist: 234.7147216796875
rank 3. neighbor: 561813, dist by lib: 243.98976135253906, actual dist: 243.98976135253906
rank 4. neighbor: 708177, dist by lib: 255.4603729248047, actual dist: 255.4603729248047
rank 5. neighbor: 706771, dist by lib: 256.31427001953125, actual dist: 256.31427001953125
====================================================================================================
query 2
    5 # LICENSE file in the root directory of this source tree.
----------------------------------------------------------------------------------------------------
rank 1. neighbor: 413247, dist by lib: 226.2454376220703, actual dist: 226.2454376220703
rank 2. neighbor: 413071, dist by lib: 229.28147888183594, actual dist: 229.28147888183594
rank 3. neighbor: 706838, dist by lib: 234.12176513671875, actual dist: 234.12176513671875
rank 4. neighbor: 880592, dist by lib: 234.49307250976562, actual dist: 234.49307250976562
rank 5. neighbor: 249062, dist by lib: 235.76683044433594, actual dist: 235.76683044433594
====================================================================================================
query 3
----------------------------------------------------------------------------------------------------
rank 1. neighbor: 669835, dist by lib: 175.47650146484375, actual dist: 175.47650146484375
rank 2. neighbor: 408764, dist by lib: 176.32640075683594, actual dist: 176.32640075683594
rank 3. neighbor: 408462, dist by lib: 183.32484436035156, actual dist: 183.32484436035156
rank 4. neighbor: 408855, dist by lib: 184.6726837158203, actual dist: 184.6726837158203
rank 5. neighbor: 551661, dist by lib: 186.27398681640625, actual dist: 186.27398681640625
====================================================================================================
query 4
----------------------------------------------------------------------------------------------------
rank 1. neighbor: 970797, dist by lib: 174.67111206054688, actual dist: 174.67111206054688
rank 2. neighbor: 125539, dist by lib: 175.67868041992188, actual dist: 175.67868041992188
rank 3. neighbor: 48044, dist by lib: 181.044189453125, actual dist: 181.044189453125
rank 4. neighbor: 191115, dist by lib: 186.7511749267578, actual dist: 186.7511749267578
rank 5. neighbor: 889039, dist by lib: 186.83950805664062, actual dist: 186.83950805664062
====================================================================================================
query 5
----------------------------------------------------------------------------------------------------
rank 1. neighbor: 340871, dist by lib: 214.27552795410156, actual dist: 214.27552795410156
rank 2. neighbor: 748397, dist by lib: 217.63961791992188, actual dist: 217.63961791992188
rank 3. neighbor: 748193, dist by lib: 219.43336486816406, actual dist: 219.43336486816406
rank 4. neighbor: 175336, dist by lib: 219.54270935058594, actual dist: 219.54270935058594
rank 5. neighbor: 716433, dist by lib: 229.2574920654297, actual dist: 229.2574920654297
```

**Comment by js1010** [comment]
I added aliases for dist_type (`euclidean, ip, cosine`) and warning in case nrz = True and dist_type = "l2"

and now, nrz is False by default

https://github.com/js1010/cuhnsw/pull/28/files

**Comment by Artif-Inteller** [comment]
I set `nrz = False` and it is fixed.

Thanks a lot.
